### PR TITLE
build: clean go build cache when necessary

### LINKED ISSUE #17078: make clean does not enable rebuild of geth
#### System information

Geth version: 1.8.12-unstable
Go Version: go1.10.3
OS & Version: OSX
Commit hash : b0cfd9c786e2eb804257c697b8e18dd6d595be6a

#### Expected behaviour
`make geth` should rebuild source code after `make clean`

#### Actual behaviour
`make clean` does not enable full rebuild of go-ethereum as below when using go version 1.10 or later.
```
$ make geth
build/env.sh go run build/ci.go install ./cmd/geth
>>> /usr/local/go/bin/go install -ldflags -X main.gitCommit=0a22ae55729298dcd9979184d0b3c28f48b272bf -s -v ./cmd/geth
Done building.
Run "/Users/hqueue/work/go-ethereum/build/bin/geth" to launch geth.
```

#### Steps to reproduce the behaviour
1. Install go with version 1.10 or higher.
2. git clone `go-ethereum`

2. `make geth`  This will trigger the first time full build of `geth`
3. `make clean`
4. `make geth` does not rebuild `geth`, just install it as below
```
$ make geth
build/env.sh go run build/ci.go install ./cmd/geth
>>> /usr/local/go/bin/go install -ldflags -X main.gitCommit=0a22ae55729298dcd9979184d0b3c28f48b272bf -s -v ./cmd/geth
Done building.
Run "/Users/hqueue/work/go-ethereum/build/bin/geth" to launch geth.
$
```

#### Backtrace
`go build` make use of cache from 1.10 (see https://golang.org/doc/go1.10#build) and go does not rebuild source code if there is no change.

We have to clean go build cache using `go clean -cache`.

````
$ go clean -cache
$ make geth  // Now it will rebuild code
````

A PR will be filed to perform `go clean -cache` when we trigger `make clean`.

### PR DESCRIPTION
**Pull request by hqueue**
- clean go build cache when necessary

Signed-off-by: Hyung-Kyu Choi <hqueue@users.noreply.github.com>

This fixes #17078

### CODE REVIEW

File: build/clean_go_build_cache.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+function version_gt() { test "$(printf '%s\n' "$@" | sort -V | head -n 1)" != "$1"; }
+function go_version {
+    version=$(go version)
+    echo $version
+        regex="([0-9].[0-9]+.[0-9])"
+        if [[ $version =~ $regex ]]; then 
+            echo ${BASH_REMATCH[1]}
+    fi
+}
```
[the remark this answers is not in the document]
Comment (hqueue) [reply build/clean_go_build_cache.sh]: Thanks for the suggestion. I will update the PR.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 4836b575ddb6081fa1c4b816c142e949374065dc
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by hqueue** [comment]
@hackmod PR updated. Please take another look :)

**Comment by karalabe** [comment]
I think your PR just won some legitimacy :) https://github.com/golang/go/issues/24355

**Pull request merged**